### PR TITLE
CLI: add arg - instantiate_only

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

allow `instantiate_only`

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by carmocca** [comment]
Can you elaborate why one would want this?

**Comment by Borda** [comment]
> Can you elaborate why one would want this?

at him moment you have to always run the fit, right? but if you want just to get the instances of Trainer, Model, and Data and then your own staff you need to overwrite `CLI.fit(...)` to be empty...
moreover (well it is docs) but from API is it not intuitive that with creating a CLI instance it automatically perform training

File: pytorch_lightning/utilities/cli.py [hunk 1]
```diff
@@ -237,6 +238,7 @@ def __init__(
             subclass_mode_data: Whether datamodule can be any `subclass
                 <https://jsonargparse.readthedocs.io/en/stable/#class-type-and-sub-classes>`_
                 of the given class.
+            skip_fit: Allow skip Trainer.fit(...), fr example if you want perform more custom flow
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/cli.py:241]:
```suggestion
            skip_fit: If ``True``, the CLI won't automatically call ``Trainer.fit(...)``.
```

**Review by ethanwharris — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by carmocca — CHANGES_REQUESTED** [review]
> at him moment you have to always run the fit, right? but if you want just to get the instances of Trainer, Model, and Data and then your own staff you need to overwrite `CLI.fit(...)` to be empty...
> moreover (well it is docs) but from API is it not intuitive that with creating a CLI instance it automatically perform training

This is the whole point of the CLI. With #7508 we could make it so not passing the function to run has the same effect as what this PR wants.

But this flag doesn't fit right with me when the only purpose of the CLI is to `fit` right now.

cc @mauvilsa for thoughts

**Comment by Borda** [comment]
> But this flag doesn't fit right with me when the only purpose of the CLI is to `fit` right now.

for me, the main purpose of CLI is instantiate all main actors - Trainer, Model and Data
then the fix is something extra and so I would rather have it as optional then mandatory

**Comment by mauvilsa** [comment]
I think we should not add a `skip_fit` parameter. The purpose of `LightningCLI` is to run a command. As @carmocca said, right now it is fit, but with #7508 there will be multiple commands. Having this parameter would not make much sense once #7508 is done.

**Comment by Borda** [comment]
> I think we should not add a `skip_fit` parameter. The purpose of `LightningCLI` is to run a command. As @carmocca said, right now it is fit, but with #7508 there will be multiple commands. Having this parameter would not make much sense once #7508 is done.

So you say that if don't want to run any Trainer method by default I cannot use CLI or need to overwrite the CLI.fit()?
@PyTorchLightning/core-contributors

**Comment by mauvilsa** [comment]
> > I think we should not add a `skip_fit` parameter. The purpose of `LightningCLI` is to run a command. As @carmocca said, right now it is fit, but with #7508 there will be multiple commands. Having this parameter would not make much sense once #7508 is done.
> 
> So you say that if don't want to run any Trainer method by default I cannot use CLI or need to overwrite the CLI.fit()?
> @PyTorchLightning/core-contributors

I am saying that `skip_fit` parameter conflicts with #7508. If there is a need that `LightningCLI` not run on instantiation, then there would be a `run` disable, not specific to `fit`. Anyway you always want to run something. The idea is that this something is implemented as part of a subclass of `LightningCLI`. Maybe can you explain better why use `LightningCLI` without running anything?

**Comment by Borda** [comment]
> Maybe can you explain better why use `LightningCLI` without running anything?

for example for the tune, or does extra action before fit or just print model details

so are we just about the name? so let's call it `intantiate_only`?

**Comment by tchaton** [comment]
Hey @mauvilsa @carmocca @Borda,

I understand Jirka use-case. I believe more users might want to rely on the LightningCLI only as an instantiation tool.
I believe having an `instantiate_only=False` argument might be useful in the long run.

Best,
T.C

**Comment by carmocca** [comment]
The discussion is scattered on both PRs but what I think is the best way forward is:

- This flag is not added as it would be removed with #7508
- We add support for subcommand as proposed in #7508
- #7508 allows for not passing any subcommand which means we just parse as this PR intends. A warning will be printed in this case as it's not a common use-case.
- In a follow-up PR, we add support for (some) combinations of common subcommands, for example: `fit-and-test`, and `tune-and-fit`.

**Comment by Borda** [comment]
@carmocca if your PR has the option to not run any trainer method, then it is fine and we do not need this :]

**Comment by tchaton** [comment]
@Borda @carmocca Good to close this PR then ?

**Comment by Borda** [comment]
closing this one in favor of #7508

**Pull request closed without merging**